File: mediator/mymediator.cpp
#include <iostream>

using namespace std;

class Chef;

class Waiter {
    public:
    virtual void Inform(Chef *chef, string message) const = 0;
};

class Chef {
    protected:
    Waiter *waiter;

    public:
    Chef(Waiter *waiter = nullptr) : waiter(waiter) {
    }
    void workwithwaiter(Waiter *waiter) {
        this->waiter = waiter;
    }
};

class SoupChef : public Chef {
    public:
    void Prepare() {
        string msg = "Soup Chef Preparing Soup";
        cout << msg << endl;
        this->waiter->Inform(this, "1");
    }
    void Decorate() {
        string msg = "Soup Chef Decorating Soup";
        cout << msg << endl;
        this->waiter->Inform(this, "2");
    }
};

class SandwichChef : public Chef {
    public:
    void GrillBread() {
        string msg = "Sandwich Chef Grilling the Bread";
        cout << msg << endl;
        this->waiter->Inform(this, "3");
    }
    void Assemble() {
        string msg = "Sandwich Chef Assembling the Dish";
        cout << msg << endl;
        this->waiter->Inform(this, "4");
    }
};

class OurWaiter : public Waiter {
    private:
    SoupChef *soupchef;
    SandwichChef *sandwichchef;

    public:
    OurWaiter(SoupChef *soupchef, SandwichChef *sandwichchef) :
              soupchef(soupchef), sandwichchef(sandwichchef) {
                this->soupchef->workwithwaiter(this);
                this->sandwichchef->workwithwaiter(this);
              }
    void Inform(Chef *chef, string msg) const override {
        if (msg == "1")
            sandwichchef->GrillBread();
        if (msg == "4") {
            soupchef->Decorate();
            sandwichchef->GrillBread();
        }
    }
};

void client() {
    SoupChef *soupchef = new SoupChef;
    SandwichChef *sandwichchef = new SandwichChef;

    OurWaiter *ourwaiter = new OurWaiter(soupchef, sandwichchef);
    soupchef->Prepare();
    soupchef->Decorate();

    sandwichchef->GrillBread();
    sandwichchef->Assemble();

    delete soupchef;
    delete sandwichchef;
    delete ourwaiter;
}

int main() {
    client();

    return(0);
}

File: mediator/swift/mymediator.swift
import Foundation

// 1. Mediator Protocol
protocol Mediator {
    func notify(sender: AnyObject, message: String)
}

// 2. Base Chef Class
class Chef {
    var waiter: Mediator?
    
    func setMediator(mediator: Mediator) {
        self.waiter = mediator
    }
    
    func prepare(order: String) {
        print("Chef is working on: \(order)")
    }
}

// 3. Concrete Chef Extensions
class SoupChef: Chef {
    override func prepare(order: String) {
        print("SoupChef: Simmering the \(order)...")
        waiter?.notify(sender: self, message: "\(order) is ready!")
    }
}

class SandwichChef: Chef {
    override func prepare(order: String) {
        print("SandwichChef: Assembling the \(order)...")
        waiter?.notify(sender: self, message: "\(order) is toasted and ready!")
    }
}

// 4. Concrete Mediator (The Waiter)
class OurWaiter: Mediator {
    private var soupChef: SoupChef
    private var sandwichChef: SandwichChef
    
    init(soupChef: SoupChef, sandwichChef: SandwichChef) {
        self.soupChef = soupChef
        self.sandwichChef = sandwichChef
        
        // Embed waiter in chefs
        self.soupChef.setMediator(mediator: self)
        self.sandwichChef.setMediator(mediator: self)
    }
    
    // informChef with chef and message as parameters
    func informChef(chef: Chef, message: String) {
        print("Waiter: Passing order to chef: \(message)")
        chef.prepare(order: message)
    }
    
    func notify(sender: AnyObject, message: String) {
        if sender is SoupChef {
            print("Waiter (Notification): Serving Soup: \(message)")
        } else if sender is SandwichChef {
            print("Waiter (Notification): Serving Sandwich: \(message)")
        }
    }
}

// 5. Main Execution
let sChef = SoupChef()
let swChef = SandwichChef()
let waiter = OurWaiter(soupChef: sChef, sandwichChef: swChef)

// Interactions
waiter.informChef(chef: sChef, message: "Tomato Basil")
print("---")
waiter.informChef(chef: swChef, message: "Club Sandwich")
